File: Java/141. Linked List Cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        // special case
        if(head == null){
            return false;
        }
        
        ListNode slow = head;
        ListNode fast = head.next;
        
        while(slow != fast){
            // if fast pointer reaches the end null or its next is the end null
            if(fast == null || fast.next == null){
                return false;
            }
            slow = slow.next;
            fast = fast.next.next;
        }
        
        return true;
    }
}

File: Python3/141. Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # special case 
        if not head:
            return False
        
        slow = head
        fast = head.next
        
        while slow != fast:
            # fast pointer reaches the end null or its next is the end null
            if not fast or not fast.next:
                return False
            slow = slow.next
            fast = fast.next.next
            
        return True